[mlir][math] Add expand patterns for acosh, asinh, atanh

jinchen62 opened the pull request:
- acosh
- asinh
- atanh
- cosh
- sinh

jinchen62 committed:
[mlir][math] Add Polynomial Approximation for acosh, asinh, atanh ops

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -615,10 +615,47 @@ TanhApproximation::matchAndRewrite(math::TanhOp op,
+    return broadcast(builder, value, shape);
+  };
+
+  // 1/2 * log((1 + x) / (1 - x))
```
Could you move it above the function, explaining it as a doc comment? Please do the same for other functions.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+    OpTy op, PatternRewriter &rewriter) const {
+  static_assert(
+      llvm::is_one_of<OpTy, math::SinhOp, math::CoshOp>::value,
+      "SinAndCosApproximation pattern expects math::SinhOp or math::CoshOp");
```
nit: `SinhAndCoshApproximation pattern`

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+    c = builder.create<arith::SubFOp>(a, b);
+  else
+    c = builder.create<arith::AddFOp>(a, b);
+  Value cstTwo = bcast(f32Cst(builder, 2.0));
```
It's better to multiply by 0.5 than to divide by 2, because we have hardware friendly instructions for multiplication.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+
+  // sinh: 1/2 * (exp(x) – exp(-x))
+  // cosh: 1/2 * (exp(x) + exp(-x))
+  Value a = builder.create<math::ExpOp>(operand);
```
Please use more meaningful names, like `a` which can be renamed as `exp`; similarly `b` can be renamed as `negExp`.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+  Value neg = builder.create<arith::NegFOp>(operand);
+  Value b = builder.create<math::ExpOp>(neg);
+  Value c;
+  if (isSine)
```
Can be written as `Value c = isSine ?  builder.create<...> : builder.create<...>`

jinchen62 replied:
I tried but got an error about can't convert operation * to value or something.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+
+namespace {
+
+template <bool isSine, typename OpTy>
```
There is no need to pass the template parameter `isSine`. It can be deduced later.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+  Value neg = builder.create<arith::NegFOp>(operand);
+  Value b = builder.create<math::ExpOp>(neg);
+  Value c;
+  if (isSine)
```
If you remove the template parameter, it would be something like 
`Value sinRes = llvm::is_one_of<OpTy, math::SinhOp>::value ? builder.create<...> : builder.create<...>`

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+
+namespace {
+
+template <bool isSine, typename OpTy>
```
Follow the same comments.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+    return broadcast(builder, value, shape);
+  };
+
+  // asinh: log(x + sqrt(x**2 + 1))
```
Move it to represent doc comments.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+  // acosh: log(x + sqrt(x**2 - 1))
+  Value squared = builder.create<arith::MulFOp>(operand, operand);
+  Value cstOne = bcast(f32Cst(builder, 1.0));
+  Value a;
```
Please use more meaning ful names.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+  Value squared = builder.create<arith::MulFOp>(operand, operand);
+  Value cstOne = bcast(f32Cst(builder, 1.0));
+  Value a;
+  if (isSine)
```
Can use the same ternary statement.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -1316,6 +1358,106 @@ LogicalResult SinAndCosApproximation<isSine, OpTy>::matchAndRewrite(
+
+  // asinh: log(x + sqrt(x**2 + 1))
+  // acosh: log(x + sqrt(x**2 - 1))
+  Value squared = builder.create<arith::MulFOp>(operand, operand);
```
You can use `builder.create<math::FmaOp>(operand, operand, cstOne)` to do it in one go. See here: https://mlir.llvm.org/docs/Dialects/MathOps/#mathfma-mathfmaop

pashu123 replied:
For subtraction, create a `negative constant`.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/PolynomialApproximation.cpp:
```diff
@@ -615,10 +615,47 @@ TanhApproximation::matchAndRewrite(math::TanhOp op,
+  Value add = builder.create<arith::AddFOp>(operand, cstOne);
+  Value neg = builder.create<arith::NegFOp>(operand);
+  Value sub = builder.create<arith::AddFOp>(neg, cstOne);
+  Value div = builder.create<arith::DivFOp>(add, sub);
```
The only value where this might give Undefined behaviour is 1.0. Could you add the test and check whether it's returning infinity. To guard the behaviour you need to see https://github.com/llvm/llvm-project/blob/803e03fbb7cd97461f349fb6e235592681fc1e6c/mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp#L257

jinchen62 replied:
Can we just allow it to be inf like checking inf here https://github.com/llvm/llvm-project/pull/90718/files#diff-c9f5949434728101e6655a80ad1e1e60add22e53632fd9b5ebcf6aa75df357ebR760 ?

pashu123 reviewed (commented):
Nice work! I have added some comments please address.

zjgarvey commented:
Why are these called polynomial approximations when you aren't approximating these functions with polynomials??

pashu123 reviewed (approved):
LGTM!

jinchen62 commented:
I will test iree tests before merge.

jinchen62 commented:
@pashu123 Could you review again? I removed the expand pattern of tanh op since it has polynomial approximation. And I added a pass including all math expand patterns so on IREE side we could change https://github.com/iree-org/iree/commit/06febe5daa7b4e977618d0752191c2b15aeecb75, if this pass doesn't make sense I would just add those missing expand patterns on IREE.

pashu123 commented:
> @pashu123 Could you review again? I removed the expand pattern of tanh op since it has polynomial approximation. And I added a pass including all math expand patterns so on IREE side we could change [iree-org/iree@06febe5](https://github.com/iree-org/iree/commit/06febe5daa7b4e977618d0752191c2b15aeecb75), if this pass doesn't make sense I would just add those missing expand patterns on IREE.

I suggest adding those missing patterns.

pashu123 commented on mlir/include/mlir/Dialect/Math/Transforms/Passes.h:
```diff
@@ -39,6 +42,7 @@ void populateExpandPowFPattern(RewritePatternSet &patterns);
 void populateExpandFPowIPattern(RewritePatternSet &patterns);
 void populateExpandRoundFPattern(RewritePatternSet &patterns);
 void populateExpandRoundEvenPattern(RewritePatternSet &patterns);
+void populateMathExpandPatterns(RewritePatternSet &patterns);
```
We don't need to expand every pattern. We want control over what patterns we want to expand and not.

pashu123 commented on mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:
```diff
@@ -615,3 +678,22 @@ void mlir::populateExpandFloorFPattern(RewritePatternSet &patterns) {
   patterns.add(convertRoundEvenOp);
 }
+
+void mlir::populateMathExpandPatterns(RewritePatternSet &patterns) {
```
I don't think we need this for the reasons mentioned above.

pashu123 commented on mlir/test/lib/Dialect/Math/TestExpandMath.cpp:
```diff
@@ -36,19 +36,7 @@ struct TestExpandMathPass
-  populateExpandFPowIPattern(patterns);
-  populateExpandRoundFPattern(patterns);
-  populateExpandRoundEvenPattern(patterns);
+  populateMathExpandPatterns(patterns);
```
You can explicitly add the patterns here for now.

pashu123 reviewed (commented):
Minor comments!

jinchen62 commented:
> Minor comments!

Addressed.

jinchen62 commented on mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:
I removed it because tanh op has polynomial approximation. Yes make sense, I have reverted it.

pashu123 replied:
Thanks. I feel users can either utilize expand pattern or polynomial approximation according to their need. Some of the operations like exp, log isn't supported on custom hardware and hence users fall back to polynomial approximation.

pashu123 reviewed (approved):
Thanks for putting this up. LGTM!

jinchen62 commented:
@pashu123 Could you merge it? I don’t have the permission. Thanks!

pashu123 commented:
> @pashu123 Could you merge it? I don’t have the permission. Thanks!

Done.
